### PR TITLE
Enhance PSI Constructor: Lower Peak Device Memory Usage

### LINKED ISSUE #4153: Double peak memory cost in `cast_memory_op`
### Describe the bug

In esolver_ks_pw.cpp:
```c++
    this->kspw_psi = GlobalV::device_flag == "gpu" 
                         || GlobalV::precision_flag == "single"
                         ? new psi::Psi<T, Device>(this->psi[0])
                         : reinterpret_cast<psi::Psi<T, Device>*>(this->psi);
```
the constructor of Psi used the function of `cast_memory_op`:
```c++
template <typename T_out, typename T_in>
struct cast_memory<T_out, T_in, container::DEVICE_CPU, container::DEVICE_GPU> {
    void operator()(
        T_out* arr_out,
        const T_in* arr_in,
        const size_t& size)
    {
        auto * arr = (T_in*) malloc(sizeof(T_in) * size);
        cudaErrcheck(cudaMemcpy(arr, arr_in, sizeof(T_in) * size, cudaMemcpyDeviceToHost));
        for (int ii = 0; ii < size; ii++) {
            arr_out[ii] = static_cast<T_out>(arr[ii]);
        }
        free(arr);
    }
};
```
the temporary memory of `arr` is same as Psi, which should be optimized as soon as possible.

### Expected behavior

_No response_

### To Reproduce

_No response_

### Environment

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

### PR DESCRIPTION
**Pull request by denghuilu**
### Reminder
- [x] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Close #4153 

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- Example: My changes might affect the performance of the application under certain conditions, and I have tested the impact on various scenarios...

### Any changes of core modules? (ignore if not applicable)
- Example: I have added a new virtual function in the esolver base class in order to ...

### CODE REVIEW

File: source/module_psi/psi.cpp [hunk 1]
```diff
@@ -163,11 +164,46 @@ Psi<T, Device>::Psi(const Psi<T_in, Device_in>& psi_in)
...
+                                                                psi_in.size());
+        // synchronize the memory from CPU to GPU
+        memory::synchronize_memory_op<T, Device, Device_in>()(this->ctx,
+                                                              psi_in.get_device(),
+                                                              this->psi,
+                                                              arr,
+                                                              psi_in.size());
+        free(arr);
+    }
+    else
```
Comment (dyzheng) [inline source/module_psi/psi.cpp:199]: which case will run this code, and what is expected to do?
initialize psi in CPU with input psi in GPU?

File: source/module_psi/psi.cpp [hunk 2]
```diff
@@ -163,11 +164,46 @@ Psi<T, Device>::Psi(const Psi<T_in, Device_in>& psi_in)
...
+    // Which means we need to initialize a GPU psi from a given CPU psi.
+    // We first malloc a memory in CPU, then cast the memory from T_in to T in CPU.
+    // Finally, synchronize the memory from CPU to GPU.
+    // This could help to reduce the peak memory usage of device.
+    else if (std::is_same<Device, DEVICE_GPU>::value && 
+             std::is_same<Device_in, DEVICE_CPU>::value) 
+    {
+        auto * arr = (T*) malloc(sizeof(T) * psi_in.size());
+        // cast the memory from T_in to T in CPU
+        memory::cast_memory_op<T, T_in, Device_in, Device_in>()(psi_in.get_device(),
```
Comment (dyzheng) [inline source/module_psi/psi.cpp:186]: Will this line allocate memory again or not?

**Comment by caic99** [comment]
Hi @denghuilu ,
I wonder if it is possible to offer an option to perform in-place conversion? Hence we can avoid all possible memory allocations.

**Review by dyzheng — APPROVED** [review]
LGTM

**Review by caic99 — APPROVED** [review]
@dyzheng This PR [patches](https://github.com/deepmodeling/abacus-develop/pull/4154/files#diff-2a220dc7976d62d38e6f8fa2c08112b6ff9244e1a18811219a906342eb0cdfa9R168-R174) the case of transferring data between CPU and GPU without type conversion. And in this PR only data transfer from CPU to GPU **with** type conversion requires memory, which is not our case.

**Comment by caic99** [comment]
@denghuilu However, I think it would be better put this optimization for T_in = T_out in implementation of `cast_memory_op`:
https://github.com/deepmodeling/abacus-develop/blob/bec5e64685a0255588b7ba955963571521cb0cc0/source/module_psi/kernels/cuda/memory_op.cu#L111-L145

**Comment by denghuilu** [comment]
> @denghuilu However, I think it would be better put this optimization for T_in = T_out in implementation of `cast_memory_op`:
> 
> https://github.com/deepmodeling/abacus-develop/blob/bec5e64685a0255588b7ba955963571521cb0cc0/source/module_psi/kernels/cuda/memory_op.cu#L111-L145

LGTM, I'll put a PR later

**Pull request merged**